### ISSUE TITLE
Spack install returns 0 even on failure

### ISSUE THREAD

**Issue by amaji** [LLNL/spack#22561]
**Spack version: 0.16.1**

I just noticed that if I don't use `--fail-fast` then `spack install` returns 0 even on a failure.

For example:
```
$ spack install gdal@3.2.0 ^netcdf-c@4.7.4~mpi ^hdf5@1.10.7~mpi %gcc@8.3.1 ; echo $?
[+] /scratch/mack8/a/amaji/spack-test-20210324/apps/openjdk/1.8.0_222-b10-gcc-10.2.0-a42p7n2
[+] /scratch/mack8/a/amaji/spack-test-20210324/apps/libiconv/1.16-gcc-10.2.0-zmywoh3
[+] /scratch/mack8/a/amaji/spack-test-20210324/apps/berkeley-db/18.1.40-gcc-10.2.0-eqrj2d5
...
...
==> Installing hdf-4.2.15-jddomeby47nactrtben375by26yrdg7o
==> No binary for hdf-4.2.15-jddomeby47nactrtben375by26yrdg7o found: installing from source
==> Using cached archive: /scratch/mack8/a/amaji/spack-test-20210324/downloads/_source-cache/archive/db/dbeeef525af7c2d01539906c28953f0fdab7dba603d1bc1ec4a5af60d002c459.tar.gz
==> hdf: Executing phase: 'autoreconf'
==> hdf: Executing phase: 'configure'
==> hdf: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j128'

3 errors found in build log:
     21     checking for config unknown-linux-gnu... no
     22     checking for config unknown-linux-gnu... no
     23     checking for config x86_64-linux-gnu... no
     24     checking for config x86_64-linux-gnu... no
     25     checking for config x86_64-unknown... no
     26     checking for config linux-gnu... found
  >> 27     gfortran: error: unrecognized command-line option '-V'
  >> 28     gfortran: fatal error: no input files
     29     compilation terminated.
     30     checking whether make sets $(MAKE)... (cached) yes
     31     checking for gcc... /scratch/mack8/a/amaji/spack-test-20210324/framework/spack/lib/spack/env/gcc/gcc
     32     checking whether the C compiler works... yes
     33     checking for C compiler default output file name... a.out
     34     checking for suffix of executables...

     ...

     490    
     491      132 |       mgsnatt = mgisattr(riid, name, nt, count, data, len(name))
     492          |                                                2
     493    ......
     494      155 |       mgsattr = mgisattr(riid, name, nt, count, data, len(name))
     495          |                                                1
  >> 496    Error: Type mismatch between actual argument at (1) and actual argument at (2) (CHARACTER(0)/INTEGER(4)).
     497    mv -f .deps/cnone.Tpo .deps/cnone.Plo
     498    make[3]: *** [Makefile:679: mfgrff.lo] Error 1
     499    make[3]: *** Waiting for unfinished jobs....
     500    mv -f .deps/dfutil.Tpo .deps/dfutil.Plo
     501    mv -f .deps/dfufp2if.Tpo .deps/dfufp2if.Plo
     502    mv -f .deps/cdeflate.Tpo .deps/cdeflate.Plo

See build log for details:
  /tmp/amaji/spack-stage-hdf-4.2.15-jddomeby47nactrtben375by26yrdg7o/spack-build-out.txt

[+] /scratch/mack8/a/amaji/spack-test-20210324/apps/py-numpy/1.19.4-gcc-10.2.0-ssip2is
[+] /scratch/mack8/a/amaji/spack-test-20210324/apps/libgeotiff/1.5.1-gcc-10.2.0-57qigv7
0
```

If I use `--fail-fast` then the exit code is 1 as expected. Can we please have this behavior fixed?

**Comment by adamjstewart** [comment]
This should be fixed by #21980

**Comment by amaji** [comment]
thanks @adamjstewart! i should have looked at the PR history.
what is the timeline for spack 0.16.2 or 0.17.0?

**Comment by adamjstewart** [comment]
No idea. @tgamblin @tldahlgren @becker33 @scheibelp?

**Comment by tgamblin** [comment]
0.17 is at least a month away, unless we cut something before the ECP meeting in April.  0.16.2 could happen soon if needed.

**Comment by tgamblin** [comment]
there are quite a few fixes already in the 0.16.2 project that we should probably do a release for

**Comment by amaji** [comment]
awesome! thanks @tgamblin @adamjstewart
